compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2060,6 +2060,25 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                                 }
                             }
                         }
+                        // If a character was expected and the found expression is a string literal
+                        // containing a single character, perhaps the user meant to write `'c'` to
+                        // specify a character literal (issue #92479)
+                        (ty::Char, ty::Ref(_, r, _)) if r.is_str() => {
+                            if let Ok(code) = self.tcx.sess().source_map().span_to_snippet(span) {
+                                if let Some(code) =
+                                    code.strip_prefix('"').and_then(|s| s.strip_suffix('"'))
+                                {
+                                    if code.len() == 1 {
```

Comment:
```suggestion
                                    if code.chars().count() == 1 {
```

However that is not very performant if the string literal is large (it keeps iterating over the string even if we already know it's larger than one). So maybe:

```suggestion
                                    if code.chars().nth(1).is_none() {
```